### ISSUE TITLE
Some concern about our plant height extractor

### ISSUE THREAD

**Issue by ZongyangLi** [terraref/extractors-3dscanner#9]
@max-zilla , @nickheyek 
Thanks a lot for the update of terrautils and plant height extractor, this is really a great progress for the extractor, it's much more clear for me to understand the extractor.

I review the codes in https://github.com/terraref/extractors-3dscanner/tree/master/plant_height, there are some questions I would like to discuss with you here to make sure the output of the extractor is something we are looking for.

1. In [this line](https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/terra_plant_height.py#L128) we are using function `gen_height_histogram()`, I think it should be `gen_height_histogram_for_Roman()`, I believe the parameters in the first function had expired, it will leads an error in height prediction.
2. From [this line](https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/terra_plant_height.py#L131) I realized that our output are 2 images, could you please show me an example of these 2 images? One big block to me for this extractor is how to integrate several output histograms and make the plot-level plant height histogram, I am wondering how is this image output handle my problem.
3. For each plant season, the boundary of each plot would be different for a short distance, in [my codes](https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/full_day_to_histogram.py#L538) I simply use a stupid `terra_common._y_row_s4` to handle the different season's boundary, and I notice that it can be done beautiful in our 'terrautils' in the future.
4. One last thing is about the 'ply_east', the extractor is only processing 'ply_west' now. One suggest is using the 'merged' las file, but I think it will not in this extractor because we are processing ply data.

**Comment by max-zilla** [comment]
1. Agree, using this method now.
2. This was in error - we were changing from NPY matrix outputs to GeoTIFFs, but that only made sense when NPY array corresponded to pixels. I reverted to writing NPY histograms for now because those don't translate to images easily.
3. I see what you mean, will see if I can achieve this with terrautils.
4. I assume your plydata library won't support LAS? Otherwise it will be nice to use merged eventually, but probably not for November.

**Comment by max-zilla** [comment]
@ZongyangLi so it looks like you have pseudocode in place where we can call BETY:
https://github.com/terraref/extractors-3dscanner/blob/master/plant_height/full_day_to_histogram.py#L543

The code we will use for terrautils will look something like:
https://github.com/terraref/terrautils/blob/master/terrautils/betydb.py#L215

"filter date" we can get from datasetname/foldername, YYYY-MM-DD. This will filter the plots to only return plots for that specific experiment based on time.
```
<set BETYDB_KEY environment variable to correct value>
>>> from terrautils.bety import get_site_boundaries
>>> # This query may take 20-30 seconds
>>> plots = get_site_boundaries("2016-01-01", city="Maricopa")

>>> len(plots)
350

>>> plots.keys()[:5]
[u'MAC Field Scanner Season 2 Range 36 Pass 3', u'MAC Field Scanner Season 2 Range 36 Pass 2', u'MAC Field Scanner Season 2 Range 36 Pass 1', u'MAC Field Scanner Season 2 Range 36 Pass 7', u'MAC Field Scanner Season 2 Range 36 Pass 6']

>>> plots['MAC Field Scanner Season 2 Range 36 Pass 3']
'{ "type": "Polygon", "coordinates": [ [ [ -111.974966423184, 33.075822211675998, 353.0 ], [ -111.974950032357995, 33.075822219426797, 353.0 ], [ -111.974950010260002, 33.075788983134103, 353.0 ], [ -111.974966400973997, 33.075788975384299, 353.0 ], [ -111.974966423184, 33.075822211675998, 353.0 ] ] ] }'
```

This will return dict of plot name: GeoJSON. then we can hopefully plug in.

**Comment by ZongyangLi** [comment]
Great, I will make a update in the afternoon. Hopefully I can finish the test today.

**Comment by ZongyangLi** [comment]
@max-zilla @dlebauer 

I have a few questions for this:

1. From function get_site_boundaries(), it seems to me that there are only 16 column in a row, it's not a 32 columns definition. 

2. For a particular day like 2017-07-01, there are only 164 plots records, most of ranges only have records for column 1 and column 16, columns between them seems missing.

3. The logic is really different if we would like to implement it with terrautils.
 a. The current code logic is base on existed ply file, we are now inputting points into row level plots, then integrate row level data into field level.
 b. The other logic might be: first get all the possible sites for the ply file(using function `get_sites(filter_date='', **kwargs)`?), then get boundaries for each sites, at last subsetting points to each plot.
 
If the planning steps is reasonable, I will still need Polygon for all the plots for that day. Or someone can confirm that boundaries for column for a particular day is always the same?

**Comment by dlebauer** [comment]
@ZongyangLi 

1. Hopefully you haven't found a crazy bug in this function ... it looks like you may have somehow selected only the border columns. For each Range, there should be 16 two-row plots and 32 E/W subplots

 here is an example of how the polygons work for season 4. We have a main plot (in black, two rows of plants) and subplots (pink and orange , with one row of plants each) 

![image](https://user-images.githubusercontent.com/464871/32630898-fa309c12-c563-11e7-9042-ef5e4d13416b.png)

2. if you are getting all columns for Range 1 and Range 54, then it appears that the function may have a bug related to only querying the Season 4 border row plots. There are actually 5 groups of plots (all plots, then divided into border rows, and short/med/tall). Here is a list of the experiments for Season 4: https://terraref.ncsa.illinois.edu/bety/experiments?search=MAC+Season+4
Currently we are keeping all of these but in the future I could see choosing one or another approach. 

And this is the API call that should return all of the polygons for Season 4:

https://terraref.ncsa.illinois.edu/bety/api/beta/experiments?key=9999999999999999999999999999999999999999&limit=none&name=MAC Season 4: All BAP Accessions&associations_mode=full_info

I can confirm that the boundaries for a column are always the same for a particular day (indeed, for a particular season).

**Comment by ZongyangLi** [comment]
@max-zilla 
I am using the example method querying the database: `get_site_boundaries(str_date, city="Maricopa")`

I can accept if all the boundaries are always the same. But do you have any idea I could query all the column and row boundaries in a more better way?

**Comment by dlebauer** [comment]
@ZongyangLi could you clarify ... what does this currently return, and what would you consider better?

**Comment by ZongyangLi** [comment]
@dlebauer The currently returns are boundaries in following sites:
MAC Field Scanner Season 4 Range 26 Column 1

MAC Field Scanner Season 4 Range 47 Column 1

MAC Field Scanner Season 4 Range 7 Column 1

MAC Field Scanner Season 4 Range 1 Column 5

MAC Field Scanner Season 4 Range 1 Column 4

MAC Field Scanner Season 4 Range 1 Column 7

MAC Field Scanner Season 4 Range 1 Column 6

MAC Field Scanner Season 4 Range 1 Column 1

MAC Field Scanner Season 4 Range 1 Column 3

MAC Field Scanner Season 4 Range 1 Column 2

MAC Field Scanner Season 4 Range 42 Column 1

MAC Field Scanner Season 4 Range 1 Column 9

MAC Field Scanner Season 4 Range 1 Column 8

MAC Field Scanner Season 4 Range 27 Column 1

MAC Field Scanner Season 4 Range 20 Column 16

MAC Field Scanner Season 4 Range 45 Column 1

MAC Field Scanner Season 4 Range 6 Column 1

MAC Field Scanner Season 4 Range 47 Column 16

MAC Field Scanner Season 4 Range 38 Column 16

MAC Field Scanner Season 4 Range 14 Column 1

MAC Field Scanner Season 4 Range 45 Column 16

MAC Field Scanner Season 4 Range 54 Column 8

MAC Field Scanner Season 4 Range 54 Column 9

MAC Field Scanner Season 4 Range 44 Column 1

MAC Field Scanner Season 4 Range 38 Column 1

MAC Field Scanner Season 4 Range 54 Column 2

MAC Field Scanner Season 4 Range 54 Column 3

MAC Field Scanner Season 4 Range 54 Column 1

MAC Field Scanner Season 4 Range 54 Column 6

MAC Field Scanner Season 4 Range 54 Column 7

MAC Field Scanner Season 4 Range 54 Column 4

MAC Field Scanner Season 4 Range 54 Column 5

MAC Field Scanner Season 4 Range 53 Column 16

MAC Field Scanner Season 4 Range 53 Column 15

MAC Field Scanner Season 4 Range 53 Column 14

MAC Field Scanner Season 4 Range 53 Column 13

MAC Field Scanner Season 4 Range 53 Column 12

MAC Field Scanner Season 4 Range 53 Column 11

MAC Field Scanner Season 4 Range 53 Column 10

MAC Field Scanner Season 4 Range 17 Column 16

MAC Field Scanner Season 4 Range 8 Column 1

MAC Field Scanner Season 4 Range 13 Column 1

MAC Field Scanner Season 4 Range 1 Column 11

MAC Field Scanner Season 4 Range 1 Column 10

MAC Field Scanner Season 4 Range 1 Column 13

MAC Field Scanner Season 4 Range 1 Column 12

MAC Field Scanner Season 4 Range 1 Column 15

MAC Field Scanner Season 4 Range 1 Column 14

MAC Field Scanner Season 4 Range 1 Column 16

MAC Field Scanner Season 4 Range 46 Column 1

MAC Field Scanner Season 4 Range 48 Column 1

MAC Field Scanner Season 4 Range 9 Column 1

MAC Field Scanner Season 4 Range 21 Column 1

MAC Field Scanner Season 4 Range 40 Column 16

MAC Field Scanner Season 4 Range 5 Column 16

MAC Field Scanner Season 4 Range 24 Column 1

MAC Field Scanner Season 4 Range 30 Column 1

MAC Field Scanner Season 4 Range 19 Column 16

MAC Field Scanner Season 4 Range 34 Column 1

MAC Field Scanner Season 4 Range 23 Column 1

MAC Field Scanner Season 4 Range 53 Column 4

MAC Field Scanner Season 4 Range 50 Column 16

MAC Field Scanner Season 4 Range 52 Column 16

MAC Field Scanner Season 4 Range 53 Column 3

MAC Field Scanner Season 4 Range 53 Column 2

MAC Field Scanner Season 4 Range 53 Column 1

MAC Field Scanner Season 4 Range 16 Column 16

MAC Field Scanner Season 4 Range 53 Column 6

MAC Field Scanner Season 4 Range 22 Column 1

MAC Field Scanner Season 4 Range 4 Column 1

MAC Field Scanner Season 4 Range 53 Column 9

MAC Field Scanner Season 4 Range 53 Column 8

MAC Field Scanner Season 4 Range 31 Column 16

MAC Field Scanner Season 4 Range 46 Column 16

MAC Field Scanner Season 4 Range 9 Column 16

MAC Field Scanner Season 4 Range 14 Column 16

MAC Field Scanner Season 4 Range 54 Column 10

MAC Field Scanner Season 4 Range 54 Column 11

MAC Field Scanner Season 4 Range 54 Column 12

MAC Field Scanner Season 4 Range 54 Column 13

MAC Field Scanner Season 4 Range 20 Column 1

MAC Field Scanner Season 4 Range 54 Column 15

MAC Field Scanner Season 4 Range 54 Column 16

MAC Field Scanner Season 4 Range 15 Column 16

MAC Field Scanner Season 4 Range 29 Column 16

MAC Field Scanner Season 4 Range 2 Column 8

MAC Field Scanner Season 4 Range 2 Column 9

MAC Field Scanner Season 4 Range 28 Column 16

MAC Field Scanner Season 4 Range 49 Column 16

MAC Field Scanner Season 4 Range 24 Column 16

MAC Field Scanner Season 4 Range 2 Column 1

MAC Field Scanner Season 4 Range 2 Column 2

MAC Field Scanner Season 4 Range 2 Column 3

MAC Field Scanner Season 4 Range 2 Column 4

MAC Field Scanner Season 4 Range 2 Column 5

MAC Field Scanner Season 4 Range 2 Column 6

MAC Field Scanner Season 4 Range 2 Column 7

MAC Field Scanner Season 4 Range 16 Column 1

MAC Field Scanner Season 4 Range 12 Column 1

MAC Field Scanner Season 4 Range 29 Column 1

MAC Field Scanner Season 4 Range 28 Column 1

MAC Field Scanner Season 4 Range 22 Column 16

MAC Field Scanner Season 4 Range 43 Column 1

MAC Field Scanner Season 4 Range 26 Column 16

MAC Field Scanner Season 4 Range 35 Column 1

MAC Field Scanner Season 4 Range 10 Column 16

MAC Field Scanner Season 4 Range 3 Column 16

MAC Field Scanner Season 4 Range 31 Column 1

MAC Field Scanner Season 4 Range 39 Column 1

MAC Field Scanner Season 4 Range 11 Column 16

MAC Field Scanner Season 4 Range 30 Column 16

MAC Field Scanner Season 4 Range 51 Column 16

MAC Field Scanner Season 4 Range 8 Column 16

MAC Field Scanner Season 4 Range 18 Column 1

MAC Field Scanner Season 4 Range 33 Column 1

MAC Field Scanner Season 4 Range 53 Column 5

MAC Field Scanner Season 4 Range 51 Column 1

MAC Field Scanner Season 4 Range 32 Column 16

MAC Field Scanner Season 4 Range 36 Column 1

MAC Field Scanner Season 4 Range 37 Column 1

MAC Field Scanner Season 4 Range 50 Column 1

MAC Field Scanner Season 4 Range 5 Column 1

MAC Field Scanner Season 4 Range 33 Column 16

MAC Field Scanner Season 4 Range 25 Column 16

MAC Field Scanner Season 4 Range 27 Column 16

MAC Field Scanner Season 4 Range 40 Column 1

MAC Field Scanner Season 4 Range 42 Column 16

MAC Field Scanner Season 4 Range 6 Column 16

MAC Field Scanner Season 4 Range 54 Column 14

MAC Field Scanner Season 4 Range 3 Column 1

MAC Field Scanner Season 4 Range 53 Column 7

MAC Field Scanner Season 4 Range 48 Column 16

MAC Field Scanner Season 4 Range 41 Column 1

MAC Field Scanner Season 4 Range 44 Column 16

MAC Field Scanner Season 4 Range 25 Column 1

MAC Field Scanner Season 4 Range 19 Column 1

MAC Field Scanner Season 4 Range 21 Column 16

MAC Field Scanner Season 4 Range 34 Column 16

MAC Field Scanner Season 4 Range 7 Column 16

MAC Field Scanner Season 4 Range 36 Column 16

MAC Field Scanner Season 4 Range 17 Column 1

MAC Field Scanner Season 4 Range 32 Column 1

MAC Field Scanner Season 4 Range 23 Column 16

MAC Field Scanner Season 4 Range 37 Column 16

MAC Field Scanner Season 4 Range 12 Column 16

MAC Field Scanner Season 4 Range 13 Column 16

MAC Field Scanner Season 4 Range 35 Column 16

MAC Field Scanner Season 4 Range 15 Column 1

MAC Field Scanner Season 4 Range 39 Column 16

MAC Field Scanner Season 4 Range 49 Column 1

MAC Field Scanner Season 4 Range 43 Column 16

MAC Field Scanner Season 4 Range 4 Column 16

MAC Field Scanner Season 4 Range 52 Column 1

MAC Field Scanner Season 4 Range 10 Column 1

MAC Field Scanner Season 4 Range 11 Column 1

MAC Field Scanner Season 4 Range 41 Column 16

MAC Field Scanner Season 4 Range 18 Column 16

MAC Field Scanner Season 4 Range 2 Column 12

MAC Field Scanner Season 4 Range 2 Column 13

MAC Field Scanner Season 4 Range 2 Column 10

MAC Field Scanner Season 4 Range 2 Column 11

MAC Field Scanner Season 4 Range 2 Column 16

MAC Field Scanner Season 4 Range 2 Column 14

MAC Field Scanner Season 4 Range 2 Column 15

As @max-zilla said, they seems like it returns only the borders.

Besides, the value for each record are something like:
str: { "type": "MultiPolygon", "coordinates": [ [ [ [ -111.975055710801669, 33.074563375249454, 0.0 ], [ -111.975039341822651, 33.074563380583719, 0.0 ], [ -111.975039327291483, 33.07453181287724, 0.0 ], [ -111.975055696371754, 33.074531807542257, 0.0 ], [ -111.975055710801669, 33.074563375249454, 0.0 ] ] ] ] }
There are 5 points in a record, what's the order of these 5 points?

**Comment by ZongyangLi** [comment]
I am guessing these 5 points are right-bottom, right-top, left-top, left-bottom, right-bottom. All of them are different except the first and the last point. So I guess I need to get all the boundaries rather than only borders.

**Comment by dlebauer** [comment]
Yep, the first and last point are duplicates. I am not sure why this is required but it is part of the geojson specification. 

What you mean by boundaries vs. borders? These are all polygons ... are you looking for the straight lines that define the columns and ranges? We don't store those in the database... but a PostGIS query could probably return them.

**Comment by ZongyangLi** [comment]
@dlebauer If you look into the site records, only range 1, 2, 53, 54 have all 16 column. Other ranges only have column 1 and 16. That's what I mean only borders.

Besides, as @max-zilla mentioned there are east/west for each plot, then there will be 32 plots in a row. But I can not find any information separating one plot to two.

**Comment by dlebauer** [comment]
What date are you using for str_date?

**Comment by ZongyangLi** [comment]
I tried 2017-06-30 and 2017-07-01

**Comment by dlebauer** [comment]
This sounds like a bug ... I would expect it to return the equivalent of 

```sql
SELECT DISTINCT sitename,
	st_astext (geometry) AS geometry
FROM
	sites
WHERE
	ID IN (
		SELECT DISTINCT
			site_id
		FROM
			experiments_sites
		WHERE
			experiment_id IN (
				SELECT
					ID
				FROM
					experiments
				WHERE
					start_date < '2017-06-30'
				AND end_date > '2017-06-30'
	/*			AND city = 'Maricopa' omitted for query speed */
			)
	)
```

Results are here: https://gist.github.com/dlebauer/f7ecdbee9f976fc43e3e096891d79a62

**Comment by dlebauer** [comment]
@ZongyangLi if you can get started with those results, I'll check with @jterstriep on why the function is not returning the expected results

**Comment by ZongyangLi** [comment]
Sure, I can start with these, assuming that all the records are there.

**Comment by dlebauer** [comment]
yep, al 2592 the records (864 two-row plots and 1728 one-row subplots) are there; Jeff is looking into the bug. 

I've reported the bug here: https://github.com/terraref/computing-pipeline/issues/372

**Comment by max-zilla** [comment]
https://github.com/terraref/terrautils/pull/39

I believe this PR fixes the issue:
```
from terrautils.betydb import get_sites
>>> s = get_sites('2017-07-01', city='Maricopa')
>>> len(s)
864
```

Note that based on previous discussions between @dlebauer and myself, we have been omitting the E/W variants from the plot lists:
```
 for t in exp['sites']:
                                # TODO: Eventually find better solution for S4 half-plots
                                if not (exp['name'].find("Season 4") > -1 and
                                            (t['site']["sitename"].endswith(" W") or
                                                 t['site']["sitename"].endswith(" E"))):
                                    if t['site'] not in results:
                                        results.append(t['site'])
```

**Comment by ZongyangLi** [comment]
@max-zilla 
I believe it works now.

Could you please share me the codes about parsing the string: str: { "type": "MultiPolygon", "coordinates": [ [ [ [ -111.975056123674676, 33.07546192772034, 0.0 ], [ -111.975039754422113, 33.07546193305857, 0.0 ], [ -111.975039739882234, 33.075430365356596, 0.0 ], [ -111.975056109128971, 33.075430360018458, 0.0 ], [ -111.975056123674676, 33.07546192772034, 0.0 ] ] ] ] }

@dlebauer , I guess I still need to add some codes inserting result into betydb, not all the height histogram, but a predict height for that plot right? 
If that is true, @rmgarnett could you please share your height prediction formula with me?

**Comment by max-zilla** [comment]
@ZongyangLi that is a geojson string defining the plot, as @dlebauer mentioned they make you repeat the first point again at the end to represent closing the polygon.

```
>>> import json
>>> from terrautils.betydb import get_site_boundaries

>>> s = get_site_boundaries('2017-07-01', city='Maricopa')
>>> g = s['MAC Field Scanner Season 4 Range 42 Column 9']
>>> g
'{ "type": "MultiPolygon", "coordinates": [ [ [ [ -111.974925433129982, 33.076037043878614, 0.0 ], [ -111.974909063770426, 33.076037049201666, 0.0 ], [ -111.974909049271801, 33.076005481501689, 0.0 ], [ -111.974925418625503, 33.076005476178743, 0.0 ], [ -111.974925433129982, 33.076037043878614, 0.0 ] ] ] ] }'
>>> j = json.loads(g)

# Upper-left
>>> j['coordinates'][0][0][0]
[-111.97492543312998, 33.076037043878614, 0.0]
# Upper-right
>>> j['coordinates'][0][0][1]
[-111.97490906377043, 33.076037049201666, 0.0]
# Lower-right
>>> j['coordinates'][0][0][2]
[-111.9749090492718, 33.07600548150169, 0.0]
# Lower-left
>>> j['coordinates'][0][0][3]
[-111.9749254186255, 33.07600547617874, 0.0]
```
...so you can use these to determine the left/right/top/bottom edges. not sure why bety nests the coordinates in a couple one-item lists (the [0][0]) but this should work.

**Comment by ZongyangLi** [comment]
@max-zilla Thanks a lot!

**Comment by rmgarnett** [comment]
@ZongyangLi: has the processing pipeline changed at all since the data you last sent me? If not, I will be happy to extract a new height formula for the current pipeline.

**Comment by ZongyangLi** [comment]
@rmgarnett 
For now all the updates are:
1. Instead of using a hard coded file to get the plot boundaries definition, use a Betydb query commend to get all plot boundaries.
2. From @max-zilla , we have been omitting the E/W variants from the plot lists, so only 16 columns for a row now, not 32 columns.

So I think I am looking for a module parameter that you had generated for estimating the plant height from height histogram.

**Comment by rmgarnett** [comment]
I hope a solution can be found for the E/W subplots; we are making manual measurements at this resolution.

**Comment by ZongyangLi** [comment]
@rmgarnett 
Sure, I think that will ask for a small update from @max-zilla to give us a E/W subplots boundaries.

**Comment by ZongyangLi** [comment]
@max-zilla 
I know that 864 plots has to be remain at this time.
Could you create another interface for me that returns E/W subplots?

**Comment by max-zilla** [comment]
@ZongyangLi https://github.com/terraref/terrautils/pull/40

I just merged this - it adds an "include_halves" flag to get_sites that will include E/W plots (false by default). This should work until we have a better solution.

**Comment by rmgarnett** [comment]
Height predictors with the best correlation to hand measurements from S4 are:

using east and west sensor:

[(35th percentile from E + 35th percentile from W) / 2] cm * 0.97841 + 25.678cm

using E sensor only:

[37th percentile from E sensor] cm * 0.94323 + 26.41cm

using W sensor only:

[41st percentile from W sensor] cm * 0.98132 + 24.852cm

**Comment by ZongyangLi** [comment]
@dlebauer 
I am getting to the end of this extractor. An example output csv I'm going to insert into BETYdb is here: https://drive.google.com/open?id=1OrcTB7bCR8iv4yoyK9kIx89J2j0k8Xbz

Please let me know if all the fields in the csv file is correct, for example, whether it should be 'height'. Besides, would you like me to insert all 3 different height estimates into BETYdb? What's traits name should it be?

@max-zilla Is there any method I can use to use the plot id getting the site name from database?
For example, what's the site name of plot 1275(in a E/W plots definition)?

**Comment by dlebauer** [comment]
@ZongyangLi the fields are good ... the trait name should be 'canopy_height' https://terraref.ncsa.illinois.edu/bety/variables/6000000007

What do you mean by 'all 3 different height estimates'? What is the difference between the three estimates (the sensor? the algorithm? other?)

One problem that I notice in the example file is that there appear to only be three unique values of height, all with the same date-time stamp, and all in the same plot. I would imagine that on each day there might be one estimate of canopy height per sensor or algorithm. I think it would be best to use the time of the image capture or laser scan, but an alternative would be to just provide the date.

**Comment by dlebauer** [comment]
@ZongyangLi if by 'all three different height estimates' you mean one per equation that Roman provided, then ... no. If anything this should be treated as a single regression with different parameters for the different cases (e.g. the equation for E plots applies to E plots and etc; we wouldn't apply the equation for a W plot to an E plot, or vice versa).

However, at least in the longer term I am curious ( @rmgarnett ) why there are three different equations. Is there a spatial process at the field scale that could be modeled (e.g. by including longitude as a covariate?) Is the difference meaningful biologically? (is the error around the prediction of a single equation smaller than the difference between equations)?

**Comment by rmgarnett** [comment]
@dlebauer Sorry, to clarify, the models are not for E/W subplots but rather for data coming from the E/W-facing laser sensors. Some plots near the E/W borders are only observed by one sensor, and thus necessitate a regression that can use just one height histogram.

**Comment by ZongyangLi** [comment]
@dlebauer 
Okay, I will change the trait name to 'canopy_height', I just want to make sure it's different from hand measurement height trait name.

Reason for the same date-time stamp is this is an integrated output comes from several scans in different time. If we just provide the date, what format should it be? 

Reason for the same plot is I didn't find a best way of generating site name, so I asked @max-zilla if I can get a E/W plots output by plot number from terrautils. If there is no other method from terrautils or betydb query, I will try to figure it out today.

And for three equations, I am sorry I didn't explain it clear before. So, are we going to use east and west facing sensor equations now:
`[(35th percentile from E + 35th percentile from W) / 2] cm * 0.97841 + 25.678cm`

**Comment by rmgarnett** [comment]
Okay, with the caveat that this will not estimate height for the full field.

**Comment by max-zilla** [comment]
@ZongyangLi you can use a URL like this:
https://terraref.ncsa.illinois.edu/bety/api/beta/sites?key=9999999999999999999999999999999999999999&id=6000010475

...but I don't believe this is supported by terrautils yet, and I don't think that's the ID number you actually have, just a bety id.

**Comment by ZongyangLi** [comment]
@dlebauer 
Here is the final example for laser height extractor: [https://drive.google.com/open?id=1N-W-4l_yGvhTFzMwmmgDBJyppIgpllm_](https://drive.google.com/open?id=1N-W-4l_yGvhTFzMwmmgDBJyppIgpllm_)

Since I don't receive comment from you, date-time stamp is still the same, site name is created by myself using 1728 plots definition without betydb query, and I am using both sensor equation: `[(35th percentile from E + 35th percentile from W) / 2] cm * 0.97841 + 25.678cm`.

@max-zilla I am going to upload my codes with a bash file now, please have a test.

**Comment by dlebauer** [comment]
@ZongyangLi sorry I didn't reply earlier ... if you just use the date the format would be "YYYY-MM-DD". 

No need to worry about this now, but for the future we could use the measurement error and growth rate to determine whether having sub-daily time precision would be useful / meaningful.

**Comment by dlebauer** [comment]
@ZongyangLi I've updated the method with the equation https://terraref.ncsa.illinois.edu/bety/methods/6000000009

**Comment by max-zilla** [comment]
@ZongyangLi so we successfully ran the extractor and generated a CSV. here are the first few lines:
```
local_datetime,canopy_height,access_level,species,site,citation_author,citation_year,citation_title,method
2017-06-01T12:00:00,1.0884285,2,Sorghum bicolor,MAC Field Scanner Season 4 Range 41 Column 2 W,"Zongyang, Li",2017,Maricopa Field Station Data and Metadata,Scanner 3d ply data to height
2017-06-01T12:00:00,0.9318829,2,Sorghum bicolor,MAC Field Scanner Season 4 Range 41 Column 3 W,"Zongyang, Li",2017,Maricopa Field Station Data and Metadata,Scanner 3d ply data to height
2017-06-01T12:00:00,0.9905875,2,Sorghum bicolor,MAC Field Scanner Season 4 Range 41 Column 3 E,"Zongyang, Li",2017,Maricopa Field Station Data and Metadata,Scanner 3d ply data to height
2017-06-01T12:00:00,0.8536101,2,Sorghum bicolor,MAC Field Scanner Season 4 Range 41 Column 4 W,"Zongyang, Li",2017,Maricopa Field Station Data and Metadata,Scanner 3d ply data to height
2017-06-01T12:00:00,0.8927465,2,Sorghum bicolor,MAC Field Scanner Season 4 Range 41 Column 4 E,"Zongyang, Li",2017,Maricopa Field Station Data and Metadata,Scanner 3d ply data to height
```

However, the upload to BETY is returning a 400 Bad Request:
```
>> r = requests.post("https://terraref.ncsa.illinois.edu/bety/api/beta/traits.csv", params={'key':'BETY_KEY'}, data=file(csv, 'rb').read(), headers={'Content-type': 'text/csv'})
```

@dlebauer @gsrohde do you see anything obviously wrong with this CSV? "canopy_height" is the trait in question. @craig-willis and I aren't as familiar with BETY api expectations.

**Comment by dlebauer** [comment]
not sure why the error isn't more useful, but there is no matching citation b/c year 2017 is wrong. Change the citation_year to 2016 and it should work. The citation_year doesn't have to reflect the date of collection

After changing the year to 2016, the following worked on a test copy of the database:

```sh 
curl -X POST --data-binary @test.csv -H "Content-Type: text/csv" "http://...betyurl.../api/v1/traits.csv?key=THEKEY"
```

**Comment by dlebauer** [comment]
where test.csv contained `cat` data from https://github.com/terraref/extractors-3dscanner/issues/9#issuecomment-345312546 ` | sed 's/2017/2016/g' > test.csv`

**Comment by max-zilla** [comment]
@dlebauer thanks!!

**Comment by gsrohde** [comment]
It may be that the python wrapper throws away textual content of error feedback.  The post, in addition to changing the database (if it succeeds), should return some JSON text containing a list of ids of new traits (if the insert succeeds) or error information (if it doesn't).

**Comment by craig-willis** [comment]
I've kicked off the process on another full day to test end-to-end.  If all looks good, we'll start the larger batch process.

**Comment by dlebauer** [comment]
Regarding https://github.com/terraref/extractors-3dscanner/issues/9#issuecomment-345336091 the curl error response was indeed sufficiently informative ... (if a bit overzealous with the `\`'s)

```sh
{
...
  },
  "errors": "{:lookup_errors=>[\"No citation could be found matching {\\\"author\\\"=>\\\"Zongyang, Li\\\", \\\"year\\\"=>\\\"2017\\\", \\\"title\\\"=>\\\"Maricopa Field Station Data and Metadata\\\"}\",
```

[Requests documentation](http://docs.python-requests.org/en/latest/user/advanced/?highlight=print maybe) suggests something like

```python
print (r.text)
```

**Comment by dlebauer** [comment]
I see in terrautils there are a lot of 

```python
json_data = r.json()
```

is it just a matter of adding this to the requests.post?

**Comment by craig-willis** [comment]
@ZongyangLi  @dlebauer 
The first end-to-end test ran over the weekend on ROGER for 2017-06-02. Please check that the Level_1 and trait data are as expected.

Data is on ROGER:
``
/projects/arpae/terraref/sites/ua-mac/Level_1/canopy_height/
``

**Comment by dlebauer** [comment]
* in the Level 1 data, I see 359 unique dates under scanner3DTop and 63 unique dates under scanner3DTop_mergedlas
* in BETYdb, there appear to be some very small and negative values from May 30 through June 2 2017

![image](https://user-images.githubusercontent.com/464871/33506416-62bb4afe-d6b5-11e7-8e1f-5a7553367cc1.png)

**Comment by craig-willis** [comment]
Moving review to
https://github.com/terraref/reference-data/issues/210

**Comment by dlebauer** [comment]
@ZongyangLi can you review this issue and close if appropriate?

**Comment by tcnichol** [comment]
Pull request that uses the new method for histogram

https://github.com/terraref/extractors-3dscanner/pull/30

does need work/feedback

**Comment by dlebauer** [comment]
closed by #30
